### DISCUSSION TITLE
`NodeNext` requirement is a breaking change in 6.2.9

### DISCUSSION THREAD

**Discussion by bfeist** [mikro-orm/mikro-orm#5789] [General]
### Describe the bug

Requiring that the project using mikro have its tsconfig set to `"moduleResolution": "NodeNext",` is a breaking change that has my project stuck at 6.2.5. My project is not ready to switch to pure ESM and depends on cjs support. Is mikro now esm only?

### Reproduction

Put `"moduleResolution": "node",` in your `tsconfig.json` and try upgrading to mikro 6.2.9.

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.9

### Node.js version

20.11

### Operating system

linux

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
Can you provide a complete reproduction? The change should affect only how the CLI is seeing your project, and as far as I am aware, it works just fine with CJS projects. NodeNext does not mean the project is ESM, for that you need to set `type: 'module'` in your package.json. In fact, all the example projects except the vanilla JS one and the getting started guide are CJS.

**Reply by bfeist** [reply]
Thanks for switching this to a discussion and taking the time to respond. Reading your response, it looks like I was conflating ESM with the NodeNext requirement.

To explain a bit why I can't use NodeNext:
Without changing my project to ESM (with `type: 'module'`), if I make the `tsconfig.json` switch to:
```
    "target": "es2022",
    "jsx": "preserve",
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
```
I get vite errors such as 
```
src/pages/testroute.tsx:5:39 - error TS1470: The 'import.meta' meta-property is not allowed in files which will build into CommonJS output.

5       <p>Sample public Vite env val: {import.meta.env.VITE_ENV_VALUE}</p>
```
With the recommended solution to switch your project to ESM. In my case this opens up an order of magnitude more cans of worms.

Here's a discussion about the problem (with no resolution) from a few years ago: https://github.com/microsoft/TypeScript/issues/49842


This is not a Mikro issue except that Mikro is now requiring NodeNext.

Thanks again for your reply.

p.s. I don't think this should have been a patch release given the severity of the impact to any projects using Mikro.

**Reply by B4nan** [reply]
> To explain a bit why I can't use NodeNext:

But you don't have to change your project to use NodeNext, only the CLI will do that when initializing ts-node internally. That change did not have anything to do with your apps, it's only about the CLI and how it brings in support for reading TS files.

**Reply by bfeist** [reply]
Apologies, I had lost track of the original error that caused this rabbit-hole journey. The error is generated by the CLI when I run `npx mikro-orm migration:up` without changing `moduleResolution` and `module` to `NodeNext`.

Locally when I run migrate up:
`error TS5109: Option 'moduleResolution' must be set to 'NodeNext' (or left unspecified) when option 'module' is set to 'NodeNext'.`

In my CI pipeline this errors out because I migrate as part of the deployment:
![image](https://github.com/mikro-orm/mikro-orm/assets/662952/4ab33188-434e-47b5-9858-9d05f2b922d7)

So while the code itself doesn't require `NodeNext` giving me no build or runtime errors, Mikro 6.2.9 CLI is overwriting my `tsconfig.json` setting for `module` making the value `NodeNext` and refuses to continue. This has the effect of Mikro CLI forcing my project to use NodeNext which as I explained above I can't do due to other dependencies.

Do you have any suggested workarounds for this?

**Reply by bfeist** [reply]
I think i've found a workaround:

in `package.json` I added `tsConfigPath` in the `mikro-orm` section to point to a `tsconfig.orm.json` file where I extend the `tsconfig.json` file with overrides for `module` and `moduleResolution`:

```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
    "declaration": true
  }
}
```

Now the CLI doesn't throw the error. Doing further testing now.

**Reply by B4nan** [reply]
Sorry for the confusion, the problem you have is already fixed in master (via d5d7f381363ab776b3c3af04a925a5547dbcdb65) but not yet released. With that commit you won't need the `tsconfig.orm.json` file anymore. You can try the latest dev version, works fine for me.
